src/components/molecules/Visualizer/usePluginInstances.ts

```diff
@@ -0,0 +1,118 @@
+  removePluginMessageSender: (id: string) => void;
+};
+
+export default ({ alignSystem, floatingWidgets, blocks }: Props) => {
+  const [pluginInstancesMeta, setPluginInstancesMeta] = useState<PluginExtensionInstance[]>([]);
+
+  useEffect(() => {
+    const instances: PluginExtensionInstance[] = [];
+
+    if (alignSystem) {
+      Object.keys(alignSystem).forEach(zoneName => {
+        const zone = alignSystem[zoneName as keyof WidgetAlignSystem];
+        if (zone) {
+          Object.keys(zone).forEach(sectionName => {
+            const section = zone[sectionName as keyof WidgetZone];
+            if (section) {
+              Object.keys(section).forEach(areaName => {
+                const area = section[areaName as keyof WidgetSection];
+                if (area?.widgets) {
+                  area?.widgets.forEach(widget => {
+                    instances.push({
+                      id: widget.id,
+                      pluginId: widget.pluginId ?? "",
+                      name: getExtensionInstanceName(widget.pluginId ?? ""),
+                      extensionId: widget.extensionId ?? "",
+                      extensionType: "widget",
+                    });
+                  });
+                }
+              });
+            }
+          });
+        }
+      });
+    }
+
+    if (floatingWidgets) {
+      floatingWidgets.forEach(widget => {
+        instances.push({
+          id: widget.id,
+          pluginId: widget.pluginId ?? "",
+          name: getExtensionInstanceName(widget.pluginId ?? ""),
+          extensionId: widget.extensionId ?? "",
+          extensionType: "widget",
+        });
+      });
+    }
+
+    if (blocks) {
+      blocks.forEach(block => {
+        instances.push({
+          id: block.id,
+          pluginId: block.pluginId ?? "",
+          name: getExtensionInstanceName(block.pluginId ?? ""),
+          extensionId: block.extensionId ?? "",
+          extensionType: "block",
+        });
+      });
+    }
+
+    setPluginInstancesMeta(instances);
+  }, [alignSystem, floatingWidgets, blocks]);
+
+  const pluginMessageSenders = useRef<Map<string, (msg: any) => void>>(new Map());
+
+  const addPluginMessageSender = useCallback((id: string, msgSender: (msg: string) => void) => {
+    pluginMessageSenders.current?.set(id, msgSender);
+  }, []);
+
+  const removePluginMessageSender = useCallback((id: string) => {
+    pluginMessageSenders.current?.delete(id);
+  }, []);
+
+  const pluginPostMessage = useCallback((id: string, msg: any, sender: string) => {
+    const msgSender = pluginMessageSenders.current?.get(id);
+    if (!msgSender) return;
+    msgSender({ data: msg, sender });
+  }, []);
+
+  const pluginInstances = useMemo(() => {
+    return {
+      meta: pluginInstancesMeta,
+      postMessage: pluginPostMessage,
+      addPluginMessageSender,
+      removePluginMessageSender,
+    };
+  }, [pluginInstancesMeta, pluginPostMessage, addPluginMessageSender, removePluginMessageSender]);
```

Comment:
Why is this necessary instead of just returning them normally? I feel like this isn't necessary but guessing you did this for a specific reason.
But, all of the returned values are either useState or useCallback, which if they change should force a re-render.